### ISSUE TITLE
redis-commander fails to start up

### ISSUE THREAD

**Issue by gauravarora** [joeferner/redis-commander#144]
I am on node v0.11.14 with npm v2.0.0. I installed redis-commander but when I try to run it I get
```
path.existsSync is now called `fs.existsSync`.
/Users/gauravarora/.nvm/v0.11.14/lib/node_modules/redis-commander/node_modules/browserify/index.js:204
        self[key] = w[key].bind(w);
                          ^
TypeError: Cannot read property 'bind' of undefined
  at /Users/gauravarora/.nvm/v0.11.14/lib/node_modules/redis-commander/node_modules/browserify/index.js:204:27
  at Array.forEach (native)
  at module.exports (/Users/gauravarora/.nvm/v0.11.14/lib/node_modules/redis-commander/node_modules/browserify/index.js:203:41)
  at module.exports (/Users/gauravarora/.nvm/v0.11.14/lib/node_modules/redis-commander/lib/app.js:48:11)
  at startWebApp (/Users/gauravarora/.nvm/v0.11.14/lib/node_modules/redis-commander/bin/redis-commander.js:158:3)
  at /Users/gauravarora/.nvm/v0.11.14/lib/node_modules/redis-commander/bin/redis-commander.js:116:10
  at /Users/gauravarora/.nvm/v0.11.14/lib/node_modules/redis-commander/lib/util.js:135:7
  at fs.js:295:14
  at Object.oncomplete (fs.js:93:15)
```

**Comment by tsabolov** [comment]
Now it's on stable channel with node 0.12. According to error stack it's caused by browserify:

```
        if (!path.existsSync(opts.file)) {
                  ^
TypeError: undefined is not a function
  at Wrap.require (/usr/local/lib/node_modules/redis-commander/node_modules/browserify/lib/wrap.js:347:19)
  at new Wrap (/usr/local/lib/node_modules/redis-commander/node_modules/browserify/lib/wrap.js:55:10)
  at module.exports (/usr/local/lib/node_modules/redis-commander/node_modules/browserify/lib/wrap.js:15:12)
  at module.exports (/usr/local/lib/node_modules/redis-commander/node_modules/browserify/index.js:50:13)
  at module.exports (/usr/local/lib/node_modules/redis-commander/lib/app.js:48:11)
  at startWebApp (/usr/local/lib/node_modules/redis-commander/bin/redis-commander.js:158:3)
  at /usr/local/lib/node_modules/redis-commander/bin/redis-commander.js:116:10
  at /usr/local/lib/node_modules/redis-commander/lib/util.js:135:7
  at fs.js:336:14
  at FSReqWrap.oncomplete (fs.js:99:15)
```

**Comment by mariusneo** [comment]
I have the same issue. Anybody knows a workaround?

**Comment by bradcosine** [comment]
I too am having the same issue on win 8.1, any update or workaround?

**Comment by mariusneo** [comment]
I have downloaded the standard distribuition tar.gz file, ran npm install on the basedir and then  node bin/redis-comander.js and it works

**Comment by bradcosine** [comment]
Definitely an issue with some of the browserify files. I changed path.existsSync to fs.existsSync and got past the initial error. There was another little change needed in browserify\index.js:

```
C:\Users\brad\AppData\Roaming\npm\node_modules\redis-commander\node_modules\browserify\index.js:204
        self[key] = w[key].bind(w);
                          ^
TypeError: Cannot read property 'bind' of undefined
  at C:\Users\brad\AppData\Roaming\npm\node_modules\redis-commander\node_modules\browserify\index.js:204:27
  at Array.forEach (native)
  at module.exports (C:\Users\brad\AppData\Roaming\npm\node_modules\redis-commander\node_modules\browserify\index.js:203:41)
  at module.exports (C:\Users\brad\AppData\Roaming\npm\node_modules\redis-commander\lib\app.js:48:11)
  at startWebApp (C:\Users\brad\AppData\Roaming\npm\node_modules\redis-commander\bin\redis-commander.js:158:3)
  at C:\Users\brad\AppData\Roaming\npm\node_modules\redis-commander\bin\redis-commander.js:116:10
  at C:\Users\brad\AppData\Roaming\npm\node_modules\redis-commander\lib\util.js:135:7
  at fs.js:336:14
  at FSReqWrap.oncomplete (fs.js:99:15)
```
Changing that function to check if w[key] is undefined resolved the issue, and I was able to get redis-commander to run.
```
    Object.keys(EventEmitter.prototype).forEach(function (key) {
        if (w[key] === undefined)
            return;
        self[key] = w[key].bind(w);
    });
```

**Comment by bgSosh** [comment]
Same issue - just started happening for me. Since upgrade to node 0.12?

**Comment by Mirocow** [comment]
i have node-0.12.0 but this mistake already exists

**Comment by scukonick** [comment]
Have this mistake as well.

**Comment by pvanroos** [comment]
Me too.  On Mac:  Yosemite 10.10.2.   

Node v0.12.0
Npm   v2.5.1

Error text after executing 'redis-commander':

No config found.
Using default configuration.
/usr/local/lib/node_modules/redis-commander/node_modules/browserify/lib/wrap.js:347
        if (!path.existsSync(opts.file)) {
                  ^
TypeError: undefined is not a function
  at Wrap.require (/usr/local/lib/node_modules/redis-commander/node_modules/browserify/lib/wrap.js:347:19)

**Comment by bryantvs24** [comment]
no solutions?!

**Comment by bryantvs24** [comment]
no resolutions?

**Comment by bgSosh** [comment]
Looks like yet another dead NPM project :/

**Comment by remkoboschker** [comment]
working fine on win8 with node v0.10.33

**Comment by sopharat** [comment]
if (!path.existsSync(opts.file)) ----- chang to => if (!fs.existsSync(opts.file))
if (path.existsSync(opts.file)) ----- chang to => if (fs.existsSync(opts.file))

Object.keys(EventEmitter.prototype).forEach(function (key) {
        self[key] = w[key].bind(w);
    });
----- chang to =>
Object.keys(EventEmitter.prototype).forEach(function (key) {
        if (w[key] === undefined) return;
        self[key] = w[key].bind(w);
    });

this work for me :D #centos

**Comment by TiTi** [comment]
Startup failing with both nodejs v0.12.2 & iojs v1.6.3 current stable releases...
```
        if (!path.existsSync(opts.file)) {
                  ^
TypeError: undefined is not a function
```

**Comment by joseph-norman** [comment]
Another bump, same issue. Going to stop using this, outdated versions and from the looks of things dead support.

**Comment by ecasilla** [comment]
+1: 
/usr/local/lib/node_modules/redis-commander/node_modules/browserify/lib/wrap.js:347
        if (!path.existsSync(opts.file)) {
                  ^
TypeError: undefined is not a function

**Comment by loucasg** [comment]
Install nodebrew: 
**curl -L git.io/nodebrew | perl - setup**
using nodebrew install node 0.10.38: 
**nodebrew install-binary v0.10.38**
switch to 0.10.38 
**nodebrew use 0.10.38**
run redis-commander

**Comment by rickystillwell** [comment]
same issue with node 0.12.2

**Comment by jsemig** [comment]
This is not a problem with redis-commander but with one of its dependencies. 
Easy way to fix this is to upgrade your browserify dependency to a newer version. I have browserify@9.0.8, works great. Just update your package.json file in the redis-commander directory with a newer version of browserify and you should be all good to go.

**Comment by evantahler** [comment]
cool.  Here's a PR https://github.com/joeferner/redis-commander/pull/151

**Comment by evantahler** [comment]
PM Merged.

**Comment by coke** [comment]
I just did an "npm install redis-commander" which seems to have pulled in
browserify@9.0.8.

Launching redis commander post install, I still get this error:

/opt/local/lib/node_modules/redis-commander/node_modules/browserify/lib/wrap.js:347
        if (!path.existsSync(opts.file)) {
                  ^
TypeError: undefined is not a function
  at Wrap.require
(/opt/local/lib/node_modules/redis-commander/node_modules/browserify/lib/wrap.js:347:19)
  at new Wrap
(/opt/local/lib/node_modules/redis-commander/node_modules/browserify/lib/wrap.js:55:10)
  at module.exports
(/opt/local/lib/node_modules/redis-commander/node_modules/browserify/lib/wrap.js:15:12)
  at module.exports
(/opt/local/lib/node_modules/redis-commander/node_modules/browserify/index.js:50:13)
  at module.exports
(/opt/local/lib/node_modules/redis-commander/lib/app.js:48:11)
  at startWebApp
(/opt/local/lib/node_modules/redis-commander/bin/redis-commander.js:158:3)
  at
/opt/local/lib/node_modules/redis-commander/bin/redis-commander.js:116:10
  at /opt/local/lib/node_modules/redis-commander/lib/util.js:135:7
  at fs.js:336:14
  at FSReqWrap.oncomplete (fs.js:99:15)



On Tue, Apr 21, 2015 at 7:48 AM, Evan Tahler <notifications@github.com>
wrote:

> PM Merged.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joeferner/redis-commander/issues/144#issuecomment-94758667>
> .
>



-- 
Will "Coke" Coleda

**Comment by rickystillwell** [comment]
I now see:

$ redis-commander 
No config found or was invalid.
Using default configuration.
No Save: false
listening on  0.0.0.0 : 8081
Redis Connection 127.0.0.1:6379 Using Redis DB #0

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: Cannot find module 'cmdparser' from '/Users/stillwell/scripts'
    at /usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:46:17
    at process (/usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:173:43)
    at ondir (/usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:188:17)
    at load (/usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:69:43)
    at onex (/usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:92:31)
    at /usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:22:47
    at Object.oncomplete (fs.js:108:15)

**Comment by jsemig** [comment]
@coke can you double check what version of browserify got pulled in because in version 9.0.8 the file which is causing your error doesn't [even exist anymore] (https://github.com/substack/node-browserify/tree/master/lib)

**Comment by rickystillwell** [comment]
$ sudo npm install -g redis-commander
/usr/local/bin/redis-commander -> /usr/local/lib/node_modules/redis-commander/bin/redis-commander.js
redis-commander@0.3.0 /usr/local/lib/node_modules/redis-commander
├── inflection@1.2.7
├── readline-browserify@0.0.4
├── cmdparser@0.0.3
├── sf@0.1.7
├── async@0.1.22
├── optimist@0.3.7 (wordwrap@0.0.2)
├── ejs@0.8.8
├── redis@0.7.3
├── node-redis-dump@0.2.1 (underscore@1.5.2, async@0.2.10, redis@0.8.6)
├── express@2.5.11 (mime@1.2.4, qs@0.4.2, mkdirp@0.3.0, connect@1.9.2)
├── browserify-middleware@3.0.1 (ms@0.6.2, mold-source-map@0.3.0, uglify-js@2.4.20, browserify@4.2.3)
└── browserify@9.0.8 (https-browserify@0.0.0, tty-browserify@0.0.0, builtins@0.0.7, constants-browserify@0.0.1, process@0.10.1, path-browserify@0.0.0, os-browserify@0.1.2, isarray@0.0.1, inherits@2.0.1, commondir@0.0.1, string_decoder@0.10.31, stream-browserify@1.0.0, defined@0.0.0, shell-quote@0.0.1, has@1.0.0, domain-browser@1.1.4, shallow-copy@0.0.1, xtend@3.0.0, duplexer2@0.0.2, deep-equal@1.0.0, querystring-es3@0.2.1, assert@1.3.0, timers-browserify@1.4.0, punycode@1.2.4, through2@1.1.1, util@0.10.3, events@1.0.2, vm-browserify@0.0.4, parents@1.0.1, read-only-stream@1.1.1, console-browserify@1.1.0, concat-stream@1.4.8, subarg@1.0.0, readable-stream@1.1.13, http-browserify@1.7.0, url@0.10.3, labeled-stream-splicer@1.0.2, shasum@1.0.1, buffer@3.2.0, browser-resolve@1.8.2, resolve@1.1.6, glob@4.5.3, JSONStream@0.10.0, deps-sort@1.3.5, browser-pack@4.0.1, browserify-zlib@0.1.4, insert-module-globals@6.2.1, syntax-error@1.1.2, crypto-browserify@3.9.14, module-deps@3.7.7)
stillwell:scripts stillwell$ redis-commander 
No config found or was invalid.
Using default configuration.
No Save: false
listening on  0.0.0.0 : 8081
Redis Connection 127.0.0.1:6379 Using Redis DB #0

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: Cannot find module 'readline-browserify' from '/Users/stillwell/scripts'
    at /usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:46:17
    at process (/usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:173:43)
    at ondir (/usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:188:17)
    at load (/usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:69:43)
    at onex (/usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:92:31)
    at /usr/local/lib/node_modules/redis-commander/node_modules/browserify-middleware/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:22:47
    at Object.oncomplete (fs.js:108:15)
stillwell:scripts stillwell$

**Comment by jsemig** [comment]
@rickystillwell update your browserify-middleware@3.0.1 to browserify-middleware@5.0.1 in your package.json file, run npm update and you should be good to go.

**Comment by rickystillwell** [comment]
Thanks @jsemig that fixed me!

**Comment by jsemig** [comment]
No problem, created a pull request for this [#152](https://github.com/joeferner/redis-commander/pull/152)

**Comment by coke** [comment]
The bumps in the PRs are showing version 0.2.x for redis commander;
however, when I uninstall redis-commander and then install it, I'm seeing:

$ npm install redis-commander
redis-commander@0.3.0 node_modules/redis-commander
├── inflection@1.2.7
├── async@0.1.22
├── sf@0.1.7
├── ejs@0.8.8
├── redis@0.7.3
├── optimist@0.3.7 (wordwrap@0.0.2)
├── express@2.5.11 (mime@1.2.4, qs@0.4.2, mkdirp@0.3.0, connect@1.9.2)
└── browserify-middleware@3.0.1 (ms@0.6.2, mold-source-map@0.3.0,
uglify-js@2.4.20, browserify@4.2.3)

I have npm installed via macports - at this point I uninstalled npm,
installed the latest version (2.8.3), and tried to install redis-commander
again; it's still picking the 0.3.0 version, which has a higher version
number than the one in the repo, but still apparently has the older deps.




On Tue, Apr 21, 2015 at 6:18 AM, Evan Tahler <notifications@github.com>
wrote:

> cool. Here's a PR #151
> <https://github.com/joeferner/redis-commander/pull/151>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joeferner/redis-commander/issues/144#issuecomment-94733601>
> .
>



-- 
Will "Coke" Coleda

**Comment by yoheimuta** [comment]
@coke 

I was in same trouble because the latest version hosted in npm is still v0.3.0.

https://www.npmjs.com/package/redis-commander

```ruby
$ npm info redis-commander

{ name: 'redis-commander',
  'dist-tags': { latest: '0.3.0' },
...
```

My solution is manual installation until the version up of npm module.

```ruby
[15:47:11 great@Mac-158 js ]$ git clone git@github.com:joeferner/redis-commander.git
[15:48:13 great@Mac-158 redis-commander master]$ npm install
[15:51:17 great@Mac-158 redis-commander master]$ ./bin/redis-commander.js
```

**Comment by itsjustcon** [comment]
+1

I'm having the same issue under OS X Yosemite *(10.10.3)* and nodejs *(v0.12.2)*.

It looks like the `resolve` module (loaded by `browser-resolve` which is loaded by `browserify` which is loaded by `browserify-middleware` which is *finally* loaded by `redis-commander`) is attempting to asynchronously load what I presume are the static web assets from my tty current working directory. Not sure why a global package would look for assets relative to the CWD and not the package's own dir.

***TL;DR*** - Something to note is I can run `sudo npm install redis-commander -g` fine, run `redis-commander` fine (I even see the redis connection established status come thru on my tty), but the process crashes whenever I try to access `http://localhost:8081/` in my browser. This is what leads me to believe it's an issue with relative paths to the static web assets.

Either way, cloning the repo or downloading a tarball of the `0.3.0` release [here](https://github.com/joeferner/redis-commander/releases/tag/v0.3.0) and running `bin/redis-commander.js` works for me.

### To Get It Working Globally
You can add a single line to `lib/app.js`, right after the last `require` statement:
```js
...
var browserify = require('browserify-middleware');
var myUtils = require('./util');

process.chdir( path.join(__dirname, '..') );    // fix the cwd

var viewsPath = path.join(__dirname, '../web/views');
var staticPath = path.join(__dirname, '../web/static');
...
```
This will change the current working directory to the root of the repo *relative to the source files* so now you can `npm link redis-commander` and be able to `redis-commander` all over the place, regardless of the tty's *actual* current working directory.

Consider merging my Pull Request #155, I hope this fix helps you guys as much as it helped me!

**Comment by tylercollier** [comment]
@joeferner Can you cut a new release with this fix in it so I can just install directly from npm?

**Comment by joeferner** [comment]
v0.3.2 published

**Comment by tylercollier** [comment]
I just tried it and it works. Thanks!

**Comment by ttpham0111** [comment]
I seems like this is still broken in the Docker image?
